<a href="https://colab.research.google.com/github/BabaGeorge22/MAT-421/blob/main/HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

HW1 - George Tome

9.1 Base-N and Binary

In base-10 (decimal), numbers are represented by the digits 0 through 9. Each digit corresponds to a power of 10. However, numbers can be represented in any base. For instance, base-2 (binary) only uses the digits 0 and 1. This is useful for computers, which ultimately deal with two voltage states (on/off). We can convert from decimal to binary and perform arithmetic operations directly in binary as well.

In [4]:
# Convert 37 (base 10) into binary by a custom function
def decimal_to_binary(n):
    """Convert a non-negative integer n to its binary representation (string)."""
    if n == 0:
        return "0"
    bits = []
    while n > 0:
        bits.append(str(n % 2))
        n //= 2
    return "".join(reversed(bits))

decimal_num = 37
binary_str = decimal_to_binary(decimal_num)
print(f"{decimal_num} (base 10) = {binary_str} (base 2)")

37 (base 10) = 100101 (base 2)


37 (base 10) = 1 x (2^5) + 0 x (2^4) + 0 x (2^3) + 1 x (2^2) + 0 x (2^1) + 0 x (2^0) = 100101 (base 2).

In [5]:
# Add 37 and 17, then multiply them, showing both decimal and binary results
a = 37
b = 17

sum_dec = a + b
prod_dec = a * b

# Use our decimal_to_binary function
bin_a = decimal_to_binary(a)
bin_b = decimal_to_binary(b)
bin_sum = decimal_to_binary(sum_dec)
bin_prod = decimal_to_binary(prod_dec)

print(f"{a} + {b} = {sum_dec} (decimal)")
print(f"{bin_a} + {bin_b} = {bin_sum} (binary)")

print("\n----------------------------\n")

print(f"{a} * {b} = {prod_dec} (decimal)")
print(f"{bin_a} * {bin_b} = {bin_prod} (binary)")

37 + 17 = 54 (decimal)
100101 + 10001 = 110110 (binary)

----------------------------

37 * 17 = 629 (decimal)
100101 * 10001 = 1001110101 (binary)


37+17=54

In binary: 100101 + 10001 = 110110, which is 1 × (2^5) + 1 × (2^4) + 0 × (2^3) + 1 × (2^2) + 1 × (2^1) + 0 × (2^0) = 54

Multiplying in binary is analogous to decimal multiplication, just using bits.

9.2 Floating Point Numbers

Floating Point representation allocates a fixed number of bits to represent very large or very small numbers by splitting those bits into:

Sign (positive or negative),

Exponent (a power of 2),

Fraction (the digits that give the precision).

In [6]:
import sys

# Get info about the platform's float
print(sys.float_info)

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)


max is the largest representable floating-point number

min is the smallest positive normalized float

In [7]:
import sys

largest = sys.float_info.max
print("Largest float:", largest)

# Adding 1 to the largest float won't change it
overflow_test = largest + 2
print("largest + 2 =", overflow_test)
print("Is largest + 2 == largest?", overflow_test == largest)

# Doubling the largest float leads to 'inf'
double_largest = largest + largest
print("largest + largest =", double_largest)

# Underflow: 2^-1075 -> 0.0
tiny_number = 2**(-1075)
print("2^-1075 =", tiny_number)
print("Is 2^-1075 == 0.0?", tiny_number == 0.0)

Largest float: 1.7976931348623157e+308
largest + 2 = 1.7976931348623157e+308
Is largest + 2 == largest? True
largest + largest = inf
2^-1075 = 0.0
Is 2^-1075 == 0.0? True


9.3 Round-off Errors

Since floating point numbers are stored with a finite number of bits, many decimal fractions cannot be represented exactly. This leads to round-off errors where arithmetic results deviate slightly from the true value.

In [8]:
print("0.1 + 0.2 + 0.3 =", 0.1 + 0.2 + 0.3)
print("Is 0.1 + 0.2 + 0.3 == 0.6?", (0.1 + 0.2 + 0.3) == 0.6)

0.1 + 0.2 + 0.3 = 0.6000000000000001
Is 0.1 + 0.2 + 0.3 == 0.6? False


In [9]:
res_sum = 0.1 + 0.2 + 0.3
print("Result =", res_sum)
print("Rounded result:", round(res_sum, 5))
print("Rounded 0.6:", round(0.6, 5))
print("Compare with rounding:", round(res_sum, 5) == round(0.6, 5))

Result = 0.6000000000000001
Rounded result: 0.6
Rounded 0.6: 0.6
Compare with rounding: True


Floating point arithmetic can yield small errors.


Using round helps compare approximate results.

In [10]:
def add_and_subtract(iterations):
    result = 1.0
    for _ in range(iterations):
        result += 1/3
    for _ in range(iterations):
        result -= 1/3
    return result

# Show how the error grows
for i in [10, 100, 1000, 10000]:
    print(f"Iterations = {i}, final result = {add_and_subtract(i)}")

Iterations = 10, final result = 0.9999999999999991
Iterations = 100, final result = 1.0000000000000002
Iterations = 1000, final result = 1.0000000000000064
Iterations = 10000, final result = 1.0000000000001166
